source/opt/ssa_rewrite_pass.cpp

```diff
@@ -0,0 +1,594 @@
+    // because this would generate an empty Phi candidate in
+    // |pred_bb|.  When |pred_bb| is later processed, a new definition
+    // for |phi_candidate->var_id_| will be lost because
+    // |phi_candidate| will still be reached by the empty Phi.
+    //
+    // Consider:
+    //
+    //       BB %23:
+    //           %38 = Phi[%i](%int_0[%1], %39[%25])
+    //
+    //           ...
+    //
+    //       BB %25: [Starts unsealed]
+    //       %39 = Phi[%i]()
+    //       %34 = ...
+    //       OpStore %i %34    -> Currdef(%i) at %25 is %34
+    //       OpBranch %23
+    //
+    // When we first create the Phi in %38, we add an operandless Phi in
+    // %39 to hold the unknown reaching def for %i.
+    //
+    // But then, when we go to complete %39 at the end.  The reaching def
+    // for %i in %25's predecessor is %38 itself.  So we miss the fact
+    // that %25 has a def for %i that should be used.
+    //
+    // By making the argument %0, we make |phi_candidate| incomplete,
+    // which will cause it to be completed after the whole CFG has
+    // been scanned.
+    uint32_t arg_id = IsBlockSealed(pred_bb)
+                          ? GetReachingDef(phi_candidate->var_id(), pred_bb)
+                          : 0;
+    phi_candidate->phi_args().push_back(arg_id);
+
+    if (arg_id == 0) {
+      found_0_arg = true;
+    } else {
+      // If this argument is another Phi candidate, add |phi_candidate| to the
+      // list of users for the defining Phi.
+      PhiCandidate* defining_phi = GetPhiCandidate(arg_id);
+      if (defining_phi && defining_phi != phi_candidate) {
+        defining_phi->AddUser(phi_candidate->result_id());
+      }
+    }
+  }
+
+  // If we could not fill-in all the arguments of this Phi, mark it incomplete
+  // so it gets completed after the whole CFG has been processed.
+  if (found_0_arg) {
+    phi_candidate->MarkIncomplete();
+    incomplete_phis_.push(phi_candidate);
+    return phi_candidate->result_id();
+  }
+
+  // Try to remove |phi_candidate|, if it's trivial.
+  uint32_t repl_id = TryRemoveTrivialPhi(phi_candidate);
+  if (repl_id == phi_candidate->result_id()) {
+    // |phi_candidate| is complete and not trivial.  Add it to the
+    // list of Phi candidates to generate.
+    phi_candidate->MarkComplete();
+    phis_to_generate_.push_back(phi_candidate);
+  }
+
+  return repl_id;
+}
+
+uint32_t SSARewriter::GetReachingDef(uint32_t var_id, ir::BasicBlock* bb) {
+  // If |var_id| has a definition in |bb|, return it.
+  const auto& bb_it = defs_at_block_.find(bb);
+  if (bb_it != defs_at_block_.end()) {
+    const auto& current_defs = bb_it->second;
+    const auto& var_it = current_defs.find(var_id);
+    if (var_it != current_defs.end()) {
+      return var_it->second;
+    }
+  }
+
+  // Otherwise, look up the value for |var_id| in |bb|'s predecessors.
+  uint32_t val_id = 0;
+  auto& predecessors = pass_->cfg()->preds(bb->id());
+  if (!IsBlockSealed(bb)) {
```

Comment:
This would be easier to read if you just asserted that the block is sealed and started the if statement below.

Reply:
I'm trying to workaround the fact that this code will not abort if we are in a release build.  We don't really have a good internal-compiler-error service.  If assertions are disabled, this returns 0, which will cause failures even in release builds.  If there's a better way of doing this, I'll happily change this.

We should really add an ICE facility to spirv-opt...